In [1]:
import pandas as pd
import numpy as np

In [2]:
df_2004 = pd.read_stata("Individual_t104.dta")
df_2024 = pd.read_excel("usu_individual_T124.xlsx")

In [ ]:
#Primera parte

In [ ]:
# Reemplazo específico para columnas por valores númericos para igualar las bases
df_2004["trimestre"] = df_2004["trimestre"].replace({
    "1er. Trimestre": 1
})
df_2004["region"] = df_2004["region"].replace({
    "Gran Buenos Aires": 1,
    "NOA": 40,
    "NEA": 41,
    "Cuyo": 42,
    "Pampeana": 43,
    "Patagónica": 44
})
df_2004["aglomerado"] = df_2004["aglomerado"].replace({
    "Gran La Plata": 2,
    "Bahía Blanca - Cerri": 3,
    "Gran Rosario": 4,
    "Gran Santa Fe": 5,
    "Gran Paraná": 6,
    "Posadas": 7,
    "Gran Resistencia": 8,
    "Gran Mendoza": 10,
    "Corrientes": 12,
    "Gran Córdoba": 13,
    "Concordia": 14,
    "Formosa": 15,
    "Jujuy - Palpalá": 19,
    "Río Gallegos": 20,
    "Gran Catamarca": 22,
    "Salta": 23,
    "La Rioja": 25,
    "San Luis - El Chorrillo": 26,
    "Gran San Juan": 27,
    "Santa Rosa - Toay": 30,
    "Ushuaia - Río Grande": 31,
    "Ciudad de Buenos Aires": 32,
    "Partidos del GBA": 33,
    "Mar del Plata - Batán": 34,
    "Río Cuarto": 36
})
df_2004["ch03"] = df_2004["ch03"].replace({
    "Jefe": 1,
    "Cónyuge/Pareja": 2,
    "Hijo/Hijastro": 3,
    "Yerno/Nuera": 4,
    "Nieto": 5,
    "Madre/Padre": 6,
    "Suegro": 7,
    "Hermano": 8,
    "Otros familiares": 9,
    "No familiares": 10
})
df_2004["ch04"] = df_2004["ch04"].replace({
    "Varón": 1,
    "Mujer": 2
})
df_2004["ch07"] = df_2004["ch07"].replace({
    "Unido": 1,
    "Casado": 2,
    "Separado o divorciado": 3,
    "Viudo": 4,
    "Soltero": 5,
    "Ns./Nr.": 99
})
df_2004["ch08"] = df_2004["ch08"].replace({
    "Obra social (incluye PAMI)": 1,
    "Mutual/Prepaga/Servicio de emergencia": 2,
    "Planes y seguros públicos": 3,
    "No paga ni le descuentan": 4,
    "Ns./Nr.": 9,
    "Obra social y mutual/prepaga/servicio de emergencia": 12,
    "Obra social y planes y seguros públicos": 13,
    "Mutual/prepaga/servicio de emergencia/planes y seguros públi": 23,
    "Obra social, mutual/prepaga/servicio de emergencia y planes": 123
})
df_2004["ch10"] = df_2004["ch10"].replace({
    "0": 0,
    "Sí, asiste": 1,
    "No asiste, pero asistió": 2,
    "Nunca asistió": 3,
    "Ns./Nr.": 99
})
df_2004["ch11"] = df_2004["ch11"].replace({
    "0": 0,
    "Público": 1,
    "Privado": 2,
    "Ns./Nr.": 99
})
df_2004["ch12"] = df_2004["ch12"].replace({
    "0": 0,
    "Jardín/Preescolar": 1,
    "Primario": 2,
    "EGB": 3,
    "Secundario": 4,
    "Polimodal": 5,
    "Terciario": 6,
    "Universitario": 7,
    "Posgrado Universitario": 8,
    "Educación especial (discapacitado)": 9,
    "Ns./Nr.": 99
})
df_2004["ch15"] = df_2004["ch15"].replace({
    "En esta localidad": 1,
    "En otra localidad": 2,
    "En otra provincia (especificar)": 3,
    "En un país limítrofe": 4,
    "En otro país": 5,
    "Ns./Nr.": 99
})
df_2004["ch16"] = df_2004["ch16"].replace({
    "En otro país": 5,
    "En un país limítrofe": 4,
    "En esta localidad": 1,
    "En otra provincia (especificar)": 3,
    "En otra localidad de esta provincia": 2,
    "No había nacido": 6,
    "Ns./Nr.": 99
})
df_2004["nivel_ed"] = df_2004["nivel_ed"].replace({
    "Primaria Incompleta (incluye educación especial)": 1,
    "Primaria Completa": 2,
    "Secundaria Incompleta": 3,
    "Secundaria Completa": 4,
    "Superior Universitaria Incompleta": 5,
    "Superior Universitaria Completa": 6,
    "Sin instrucción": 7
})
df_2004["estado"] = df_2004["estado"].replace({
    "Entrevista individual no realizada (no respuesta al cuestion": 0,
    "Ocupado": 1,
    "Desocupado": 2,
    "Inactivo": 3,
    "Menor de 10 años": 4
})
df_2004["cat_ocup"] = df_2004["cat_ocup"].replace({
    "0": 0,
    "Patrón": 1,
    "Cuenta propia": 2,
    "Obrero o empleado": 3,
    "Trabajador familiar sin remuneración": 4
})
df_2004["cat_inac"] = df_2004["cat_inac"].replace({
    "0": 0,
    "Jubilado/pensionado": 1,
    "Rentista": 2,
    "Estudiante": 3,
    "Ama de casa": 4,
    "Menor de 6 años": 5,
    "Discapacitado": 6,
    "Otros": 7
})
df_2004["pp03c"] = df_2004["pp03c"].replace({
    "0": 0,
    "...un sólo empleo/ocupación/actividad?": 1,
    "...más de un empleo/ocupación/actividad?": 2,
    "Ns./Nr.": 99
})
df_2004["pp03h"] = df_2004["pp03h"].replace({
    "0": 0,
    "...podía trabajarlas esa semana?": 1,
    "...podía empezar a trabajarlas en dos semanas a más tardar?": 2,
    "...no podía trabajar más horas?": 3,
    "Ns./Nr.": 99
})
df_2004["intensi"] = df_2004["intensi"].replace({
    "0": 0,
    "Subocupación horaria Demandante": 1,
    "Subocupación horaria No Demandante": 2,
    "Ocupación plena": 3,
    "Sobreocupación horaria": 4,
    "Ocupado que no trabajó en la semana": 5,
    "Ns./Nr.": 99
})
df_2004["pp04a"] = df_2004["pp04a"].replace({
    "0": 0,
    "...estatal?": 1,
    "...privada?": 2,
    "...de otro tipo? (especificar)": 3,
    "Ns./Nr.": 99
})
df_2004["pp04c"] = df_2004["pp04c"].replace({
    "0": 0,
    "1 persona": 1,
    "2 personas": 2,
    "3 personas": 3,
    "4 personas": 4,
    "5 personas": 5,
    "de 6 a 10 personas": 6,
    "de 11 a 25 personas": 7,
    "de 26 a 40 personas": 8,
    "de 41 a 100 personas": 9,
    "de 101 a 200 personas": 10,
    "de 201 a 500 personas": 11,
    "más de 500 personas": 12,
    "Ns./Nr.": 99
})
df_2004["pp04g"] = df_2004["pp04g"].replace({
    "0": 0,
    "En un local/oficina/establecimiento/negocio/taller/chacra/fi": 1,
    "En puesto o kiosco fijo callejero": 2,
    "En vehículos: bicicleta/moto/auto/barco/bote (no incluye ser": 3,
    "En vehículo para transporte de personas y mercaderías-aéreo,": 4,
    "En obras en construcción, de infraestructura, minería o simi": 5,
    "En esta vivienda": 6,
    "En la vivienda del socio o del patrón": 7,
    "En el domicilio/local de los clientes": 8,
    "En la calle/espacios públicos/ambulante/de casa en casa/pues": 9,
    "En otro lugar": 10,
    "Ns./Nr.": 99
})
df_2004["pp05c_1"] = df_2004["pp05c_1"].replace({
    "0": 0,
    "Propio (del negocio)": 1,
    "Prestado/alquilado": 2,
    "No tiene": 3
})
df_2004["pp05c_2"] = df_2004["pp05c_2"].replace({
    "0": 0,
    "Propio (del negocio)": 1,
    "Prestado/alquilado": 2,
    "No tiene": 3,
    "Ns./Nr.": 99
})
df_2004["pp05c_3"] = df_2004["pp05c_3"].replace({
    "0": 0,
    "Propio (del negocio)": 1,
    "Prestado/alquilado": 2,
    "No tiene": 3,
    "Ns./Nr.": 99
})
df_2004["pp05f"] = df_2004["pp05f"].replace({
    "0": 0,
    "Un solo cliente? (persona, empresa)": 6,
    "Distintos clientes? (incluye público en general)": 7
})
df_2004["pp05h"] = df_2004["pp05h"].replace({
    "0": 0,
    "Menos de un mes": 1,
    "De 1 a 3 meses": 2,
    "Más de 3 a 6 meses": 3,
    "Más de 6 meses a 1 año": 4,
    "Más de 1 a 5 años": 5,
    "Más de 5 años": 6,
    "Ns./Nr.": 99
})
df_2004["pp06e"] = df_2004["pp06e"].replace({
    "0": 0,
    "Es una sociedad jurídicamente constituida? (S.A., S.R.L., Co": 1,
    "Es una sociedad de otra forma legal?": 2,
    "O es una sociedad convenida de palabra?": 3,
    "Ns./Nr.": 99
})
df_2004["pp07a"] = df_2004["pp07a"].replace({
    "0": 0,
    "Menos de un mes": 1,
    "1 a 3 meses": 2,
    "Más de 3 a 6 meses": 3,
    "Más de 6 a 12 meses": 4,
    "Más de 1 a 5 años": 5,
    "Más de 5 años": 6,
    "Ns./Nr.": 99
})
df_2004["pp07c"] = df_2004["pp07c"].replace({
    "0": 0,
    "Sí (incluye changa, trabajo transitorio, por tarea u obra, s": 1,
    "No (Incluye permanente, fijo, estable, de planta)": 2,
    "Ns./Nr.": 99
})
df_2004["pp07j"] = df_2004["pp07j"].replace({
    "0": 0,
    "...de día? (mañana/tarde)": 1,
    "...de noche": 2,
    "...de otro tipo? (rotativo, día y noche, guardias con franco": 3,
    "Ns./Nr.": 99
})
df_2004["pp07k"] = df_2004["pp07k"].replace({
    "0": 0,
    "...le dan recibo con sello/membrete/firma del empleador?": 1,
    "...le dan un papel/recibo sin nada?": 2,
    "...entrega una factura?": 3,
    "...no le dan ni entrega nada?": 4,
    "no cobra, es trabajador sin pago, ad-honorem": 5,
    "Ns./Nr.": 99
})
df_2004["pp09a"] = df_2004["pp09a"].replace({
    "0": 0,
    "Ciudad de Buenos Aires": 1,
    "Partidos del GBA": 2,
    "Ambos": 3,
    "En otro lugar": 4,
    "Ns./Nr.": 99,
    })
df_2004["pp10a"] = df_2004["pp10a"].replace({
    "0": 0,
    "...menos de 1 mes?": 1,
    "...de 1 a 3 meses?": 2,
    "...más de 3 a 6 meses?": 3,
    "...más de 6 a 12 meses?": 4,
    "...más de 1 año?": 5,
    "Ns./Nr.": 99
})
df_2004["pp10e"] = df_2004["pp10e"].replace({
    "0": 0,
    "...menos de 1 mes?": 1,
    "...de 1 a 3 meses?": 2,
    "...más de 3 a 6 meses?": 3,
    "...más de 6 a 12 meses?": 4,
    "...más de 1 a 3 años?": 5,
    "...más de 3 años?": 6,
    "Ns./Nr.": 99
})
df_2004["pp11a"] = df_2004["pp11a"].replace({
    "0": 0,
    "...estatal?": 1,
    "...privada?": 2,
    "...de otro tipo?": 3,
    "Ns./Nr.": 99
})
df_2004["pp11c"] = df_2004["pp11c"].replace({
    "0": 0,
    "1 persona": 1,
    "2 personas": 2,
    "3 personas": 3,
    "4 personas": 4,
    "5 personas": 5,
    "de 6 a 10 personas": 6,
    "de 11 a 25 personas": 7,
    "de 26 a 40 personas": 8,
    "de 41 a 100 personas": 9,
    "de 101 a 200 personas": 10,
    "de 201 a 500 personas": 11,
    "más de 500 personas": 12,
    "Ns./Nr.": 99
})
df_2004["pp11l"] = df_2004["pp11l"].replace({
    "0": 0,
    "Falta de clientes/clientes que no pagan": 1,
    "Falta de capital/equipamiento": 2,
    "Trabajo estacional": 3,
    "Tenía gastos demasiado altos": 4,
    "Otras causas laborales (especificar)": 5,
    "Causas personales (matrimonio, embarazo, cuidado de hijos o": 7,
    "Ns./Nr.": 99
})
df_2004["pp11l1"] = df_2004["pp11l1"].replace({
    "0": 0,
    "...una changa, trabajo transitorio, por tarea u obra, suplen": 1,
    "...un trabajo permanente, fijo estable, de planta, etc.?": 2,
    "Ns./Nr.": 99
})
df_2004["pp11m"] = df_2004["pp11m"].replace({
    "0": 0,
    "...un plan de empleo?": 1,
    "...un período de prueba?": 2,
    "...otro tipo de trabajo?": 3
})
df_2004["pp11o"] = df_2004["pp11o"].replace({
    "0": 0,
    "Despido/cierre": 1,
    "Por retiro voluntario del sector público": 2,
    "Por jubilación": 3,
    "Fin del trabajo temporario/estacional": 4,
    "Le pagaban poco/no le pagaban": 5,
    "Malas relaciones laborales/malas condiciones de trabajo (ins": 6,
    "Renuncia obligada/pactada": 7,
    "Otras causas laborales": 8,
    "Por razones personales": 9
})
df_2004["pj1_1"] = df_2004["pj1_1"].replace({
    "0": 0,
    "Tiene plan jefas y jefes": 1,
    "Ns./Nr.": 99
})
df_2004["pj2_1"] = df_2004["pj2_1"].replace({
    "0": 0,
    "Tiene plan jefas y jefes": 1,
    "Ns./Nr.": 99
})
df_2004["pj3_1"] = df_2004["pj3_1"].replace({
    "0": 0,
    "Tiene plan jefas y jefes": 1,
    "Ns./Nr.": 99
})

In [ ]:
def reemplazar_valores(df, columnas):
    reemplazos = {
        "Sí": 1,
        "SÍ": 1,
        "No": 2,
        "NO": 2,
        "Ns./Nr.": 99
    }
    df[columnas] = df[columnas].replace(reemplazos)
    return df

# Lista completa de columnas
columnas_a_reemplazar = ["pp02c6", "pp02c7", "pp02c8", "pp02h", "pp02i", "pp03g", "pp03i", "pp03j", "pp05e",
    "pp06a", "pp06h", "pp07f1", "pp07f2", "pp07f3", "pp07f4", "pp07f5", "pp07g1",
    "pp07g2", "pp07g3", "pp07g4", "pp07h", "pp07i", "pp10c", "pp10d", "pp11n", "pp11p",
    "pp11q", "pp11r", "pp11s", "pp11t",
    "h15", "ch09", "ch13", "pp02c1", "pp02c2", "pp02c3", "pp02c4", "pp02c5"
]

# Aplicar reemplazo en df_2004
df_2004 = reemplazar_valores(df_2004, columnas_a_reemplazar)

In [ ]:
#df_2004[columnas_a_reemplazar] = df_2004[columnas_a_reemplazar].astype(str)

In [ ]:
# Observamos los cambios realizados
df_2004

In [9]:
# Ponemos los títulos en mayúsculas para filtrar la región
df_2004.columns = df_2004.columns.str.upper()
df_2024.columns = df_2024.columns.str.upper()

In [10]:
# Filtrar región NOA (REGION == 40)
df_2004_noa = df_2004[df_2004["REGION"] == 40]
df_2024_noa = df_2024[df_2024["REGION"] == 40]

In [ ]:
# Mostrar conteo para verificar
print("Cantidad de registros en la región NOA:")
print("2004:", len(df_2004_noa))
print("2024:", len(df_2024_noa))

In [12]:
# Unión de las dos bases
df_unido = pd.concat([df_2004_noa, df_2024_noa], ignore_index=True)

In [13]:
# definicion de la base de datos "respondieron" para tener los datos ordenados desde el principio.
respondieron = df_unido[(df_unido["ESTADO"].notna()) & (df_unido["ESTADO"] != 0)]

In [14]:
from sklearn.model_selection import train_test_split
import statsmodels.api as sm

In [21]:
# Copia para trabajar con seguridad
respondieron = respondieron.copy()

# Crear variable dependiente: 1 si desocupado, 0 si no
respondieron["desocupado"] = (respondieron["ESTADO"] == 2).astype(int)

# Variables de interés seleccionadas
variables_interes = [
    "CH04",  # Sexo
    "CH06",  # Edad
    "ESTADO",  # Condición de actividad
    "CAT_OCUP", # Categoría ocupacional
    "AGLOMERADO", # Código de Aglomerado
    "NIVEL_ED",  # Nivel educativo
    "P47T",  # Ingreso total individual
    "P21",  # Ingreso ocupación principal
    "CH03",  # Parentesco
    "CH07",  # Estado civil
    "CH08",  # Cobertura médica
    "CH10",  # Asistencia educativa
    "CH11",  # Tipo de establecimiento educativo
    "CH12",  # Nivel cursado
    "CH13",  # Nivel finalizado
]

# Subconjunto X e y
X = respondieron[variables_interes]
y = respondieron["desocupado"]

# División train/test
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=444)

# Conversión segura a numérico
X_train = X_train.apply(pd.to_numeric, errors='coerce')
X_test = X_test.apply(pd.to_numeric, errors='coerce')

# Imputar valores faltantes con 0
X_train = X_train.fillna(0)
X_test = X_test.fillna(0)

# Agregar columna de unos
X_train = sm.add_constant(X_train, has_constant='add')
X_test = sm.add_constant(X_test, has_constant='add')

# Tabla de diferencias de medias
mean_train = X_train.mean()
mean_test = X_test.mean()
diff_means = pd.DataFrame({
    "Media Train": mean_train,
    "Media Test": mean_test,
    "Diferencia": mean_train - mean_test
})

# Mostrar tabla
print(diff_means)

             Media Train    Media Test   Diferencia
const           1.000000      1.000000     0.000000
CH04            1.520746      1.525424    -0.004678
CH06           32.202741     32.196575     0.006166
ESTADO          2.313811      2.324655    -0.010844
CAT_OCUP        1.188736      1.168793     0.019943
AGLOMERADO     17.949820     17.570680     0.379141
NIVEL_ED        3.600809      3.554779     0.046030
P47T        78375.355303  75587.628167  2787.727136
P21         48697.515503  45326.915778  3370.599725
CH03            2.750449      2.738773     0.011676
CH07            3.716896      3.745239    -0.028342
CH08            2.854329      2.833479     0.020850
CH10            1.656456      1.656124     0.000332
CH11            0.412747      0.393500     0.019247
CH12            3.713376      3.713262     0.000114
CH13            1.469518      1.473877    -0.004360


La tabla presenta las medias de las variables explicativas seleccionadas en los conjuntos de entrenamiento (train) y prueba (test), así como la diferencia entre ellas. Las diferencias que se observan son, en general, bastante pequeñas, lo que sugiere que la partición aleatoria realizada con train_test_split (usando random_state=444) fue adecuado.

Algunos puntos a resaltar:

Las variables como CH04 (sexo), CH06 (edad), CH07 (estado civil) y otras variables categóricas como CH08 (cobertura médica) y CH10 (asistencia educativa) muestran diferencias menores a 0.03 puntos, lo que indica que ambas muestras representan bien al total de la población.

En cuanto a las variables numéricas más sensibles, tenemos:

P47T (ingreso total individual): diferencia ≈ 2.788 pesos

P21 (ingreso de la ocupación principal): diferencia ≈ 3.370 pesos

Aunque estas diferencias pueden parecer un poco más grandes en términos absolutos, hay que ponerlas en perspectiva, ya que los montos promedios rondan entre 75.000 y 78.000 pesos. Por lo tanto, en términos relativos, la diferencia es inferior al 5%, lo cual sigue siendo aceptable para un muestreo aleatorio.

AGLOMERADO muestra una diferencia promedio de ≈ 0.38. Dado que esta variable es un identificador geográfico (y no ordinal), esta diferencia no tiene una interpretación directa en cuanto a magnitud o sesgo.

En resumen, no se han encontrado indicios de que el conjunto de entrenamiento esté sesgado en comparación con el conjunto de prueba, lo que nos permite avanzar con el entrenamiento del modelo con confianza en la representatividad de los datos.